File: services/python/flask/src/routes/middleware/chaos.py
from random import randrange
from gevent import sleep
from flask import current_app

from app.config import Config
from routes.errors import send_error, AppErrors


def chaos_middleware():
    config: Config = current_app.extensions["config"]

    if config.latency.enabled and randrange(0, 101) < config.latency.rate:
        sleep(randrange(500, 1501) / 1_000_000)
    if config.error.enabled and randrange(0, 101) < config.error.rate:
        return send_error(AppErrors.Internal)

    return None

File: services/typescript/hono/src/routes/middleware/chaos.ts
import type { Context, MiddlewareHandler, Next } from "hono";
import { randomInt } from "node:crypto";
import { setTimeout } from "node:timers/promises";
import type { State } from "app";
import { AppErrors, sendError } from "../errors.js";

export function chaosMiddleware(state: State): MiddlewareHandler {
  return async (c: Context, next: Next) => {
    const config = state.config;

    if (config.latency.enabled && randomInt(0, 101) < config.latency.rate) {
      // Note: can't do sub ms but logic is same
      await setTimeout(randomInt(500, 1501) / 1000);
    }
    if (config.error.enabled && randomInt(0, 101) < config.error.rate) {
      return sendError(c, AppErrors.Internal);
    }

    return await next();
  };
}
